File: sdks/dart/lib/src/changes.dart
part of 'client.dart';

final _identityOriginPattern = RegExp(r'^[0-9a-f]{32}$');
const _maxIdentityFrameBytes = 1 << 20;
const _maxIdentityChunkItems = 1024;

/// A portable per-origin cursor whose values are the NEXT expected sequence.
///
/// This differs from a durable last-applied checkpoint. An identity stream
/// never uses the responder-local mutation-log position.
final class IdentityNextCursor {
  /// Copies and validates the next sequence expected from each origin.
  IdentityNextCursor(Map<String, BigInt> nextSequences)
    : nextSequences = Map<String, BigInt>.unmodifiable(nextSequences) {
    for (final entry in this.nextSequences.entries) {
      _validateIdentityOrigin(entry.key);
      if (entry.value < BigInt.one || entry.value > _maxUint64) {
        throw _invalidArgumentException(
          'identity cursor sequence must fit uint64 and be positive',
        );
      }
    }
  }

  /// Converts a durable last-applied vector, rejecting uint64 exhaustion.
  factory IdentityNextCursor.fromLastApplied(
    Map<String, BigInt> lastSequences,
  ) {
    for (final last in lastSequences.values) {
      if (last < BigInt.zero || last >= _maxUint64) {
        throw _invalidArgumentException(
          'last-applied identity sequence cannot advance',
        );
      }
    }
    return IdentityNextCursor({
      for (final entry in lastSequences.entries)
        entry.key: entry.value + BigInt.one,
    });
  }

  /// Immutable canonical origin IDs mapped to NEXT expected sequences.
  final Map<String, BigInt> nextSequences;
}

/// One value-free identity stream frame.
sealed class IdentityFrame {
  const IdentityFrame();
}

/// The atomic publication cutoff sent first during bootstrap.
///
/// Values are LAST committed sequences. This is not a cluster-wide
/// consensus or freshness guarantee; the live tail starts at this cut.
final class IdentityCheckpointFrame extends IdentityFrame {
  IdentityCheckpointFrame._(Map<String, BigInt> lastSequences)
    : lastSequences = Map<String, BigInt>.unmodifiable(lastSequences);

  /// Immutable per-origin LAST committed sequence vector.
  final Map<String, BigInt> lastSequences;
}

/// The exact category of a committed graph mutation.
enum IdentityOperation {
  /// A Vertex Put, including a causal barrier or born-expired overwrite.
  putVertex,

  /// A Vertex Delete, including exact victims of a capped prefix Delete.
  deleteVertex,

  /// An additive Edge write.
  addEdge,

  /// An idempotent Edge replacement.
  putEdge,

  /// An Edge Delete, including exact victims of a capped prefix Delete.
  deleteEdge,
}

/// The original mutation's Hybrid Logical Clock coordinate.
final class IdentityHlc {
  const IdentityHlc._({
    required this.wallNanoseconds,
    required this.logical,
    required this.nodeId,
  });

  /// Signed nanoseconds since the Unix epoch, without DateTime precision loss.
  final BigInt wallNanoseconds;

  /// Logical counter within the physical tick.
  final int logical;

  /// Canonical origin node ID.
  final String nodeId;
}

/// One bounded fragment of a committed mutation's exact identities.
///
/// The durable per-origin cursor must advance only after [isLast] is applied.
/// A final fragment can have no identities. No graph values, Edge weights,
/// contribution IDs, or credentials are present in this type.
final class IdentityChunkFrame extends IdentityFrame {
  IdentityChunkFrame._({
    required this.origin,
    required this.sequence,
    required this.hlc,
    required this.operation,
    required this.chunkIndex,
    required this.isLast,
    required this.firstItemIndex,
    required List<String> vertexKeys,
    required List<EdgeRef> edgeKeys,
  }) : vertexKeys = List<String>.unmodifiable(vertexKeys),
       edgeKeys = List<EdgeRef>.unmodifiable(edgeKeys);

  /// Canonical 32-character hexadecimal origin ID.
  final String origin;

  /// Origin-local committed sequence, represented without signed truncation.
  final BigInt sequence;

  /// Causal coordinate of the original mutation.
  final IdentityHlc hlc;

  /// Original operation category.
  final IdentityOperation operation;

  /// Zero-based chunk index within [sequence].
  final int chunkIndex;

  /// True only for this mutation's final chunk.
  final bool isLast;

  /// First projected identity index in this chunk.
  final int firstItemIndex;

  /// Exact Vertex keys to invalidate.
  final List<String> vertexKeys;

  /// Exact Edge identities to invalidate.
  final List<EdgeRef> edgeKeys;
}

/// Value-free, deployment-scoped replication changes for cache invalidation.
extension LanternChanges on LanternClient {
  /// Opens one identity-only Subscribe stream.
  ///
  /// Set [bootstrap] for an atomic checkpoint followed by live chunks. On
  /// resume, pass a [cursor] of NEXT expected sequences; absent origins start
  /// at their first retained mutation. A `failedPrecondition` gap requires a
  /// fresh bootstrap and resident-key revalidation. No retry, endpoint
  /// discovery, OS scheduling, or tenant filtering is performed here.
  /// Unexpected clean EOF also requires bootstrap and resident-key
  /// revalidation, and is reported as `failedPrecondition`.
  ///
  /// This long-lived stream ignores the client's default unary timeout. An
  /// explicit [LanternCallOptions.timeout] or deadline still applies. Cancel
  /// the returned subscription or its cancellation token to release the RPC.
  Stream<IdentityFrame> subscribeIdentity({
    IdentityNextCursor? cursor,
    bool bootstrap = false,
    LanternCallOptions? options,
  }) {
    _ensureOpen();
    if (bootstrap && (cursor?.nextSequences.isNotEmpty ?? false)) {
      throw _invalidArgumentException(
        'identity bootstrap requires an empty cursor',
      );
    }
    final request = $replication.SubscribeRequest(
      fromSeqPerOrigin: (cursor?.nextSequences ?? const <String, BigInt>{})
          .entries
          .map((entry) => MapEntry(entry.key, _uint64ToFixnum(entry.value))),
      projection:
          $replication.SubscribeProjection.SUBSCRIBE_PROJECTION_IDENTITY_ONLY,
      bootstrap: bootstrap,
    );
    final raw = $replication_client.LanternReplicationServiceClient(
      _invoker.transport,
    );
    final stream = _invoker.invokeStream<$replication.SubscribeResponse>(
      call:
          ({
            required headers,
            required signal,
            required onHeader,
            required onTrailer,
          }) => raw.subscribe(
            request,
            headers: headers,
            signal: signal,
            onHeader: onHeader,
            onTrailer: onTrailer,
          ),
      options: LanternCallOptions(
        timeout: options?.timeout,
        deadline: options?.deadline,
        cancellation: options?.cancellation,
        retry: false,
        disableDefaultTimeout: true,
      ),
    );
    return _decodeIdentityFrames(
      stream,
      bootstrap: bootstrap,
      cancellation: options?.cancellation,
    );
  }
}

Stream<IdentityFrame> _decodeIdentityFrames(
  Stream<$replication.SubscribeResponse> source, {
  required bool bootstrap,
  LanternCancellationToken? cancellation,
}) {
  StreamSubscription<$replication.SubscribeResponse>? upstream;
  late final StreamController<IdentityFrame> controller;
  var checkpointSeen = false;
  var stopped = false;
  var paused = false;

  void fail(Object error, StackTrace stack) {
    if (stopped) return;
    stopped = true;
    controller.addError(error, stack);
    final active = upstream;
    if (active != null) active.cancel().ignore();
    unawaited(controller.close());
  }

  controller = StreamController<IdentityFrame>(
    sync: true,
    onListen: () {
      final active = source.listen(
        (response) {
          if (stopped) return;
          try {
            switch (response.whichEvent()) {
              case $replication.SubscribeResponse_Event.checkpoint:
                if (!bootstrap || checkpointSeen) {
                  throw _internalSdkException('unexpected identity checkpoint');
                }
                if (response.writeToBuffer().length > _maxIdentityFrameBytes) {
                  throw _internalSdkException(
                    'identity checkpoint exceeds size limit',
                  );
                }
                checkpointSeen = true;
                controller.add(_decodeIdentityCheckpoint(response.checkpoint));
              case $replication.SubscribeResponse_Event.identityChunk:
                if (bootstrap && !checkpointSeen) {
                  throw _internalSdkException(
                    'identity chunk preceded checkpoint',
                  );
                }
                controller.add(_decodeIdentityChunk(response));
              case $replication.SubscribeResponse_Event.mutation:
              case $replication.SubscribeResponse_Event.notSet:
                throw _internalSdkException('unexpected identity stream frame');
            }
          } catch (error, stack) {
            fail(error, stack);
          }
        },
        onError: (Object error, StackTrace stack) => fail(error, stack),
        onDone: () {
          if (stopped) return;
          if (cancellation?.isCanceled ?? false) {
            stopped = true;
            unawaited(controller.close());
          } else if (bootstrap && !checkpointSeen) {
            fail(
              _internalSdkException(
                'identity bootstrap ended without checkpoint',
              ),
              StackTrace.current,
            );
          } else {
            fail(
              LanternFailedPreconditionException._(
                _ErrorData(
                  transportCode: connect.Code.failedPrecondition.value,
                  transportCodeName: connect.Code.failedPrecondition.name,
                  message:
                      'identity stream ended unexpectedly; bootstrap and revalidate resident keys',
                  headers: {},
                  trailers: {},
                  metadata: {},
                ),
              ),
              StackTrace.current,
            );
          }
        },
      );
      upstream = active;
      if (paused) active.pause();
      if (stopped) active.cancel().ignore();
    },
    onPause: () {
      paused = true;
      upstream?.pause();
    },
    onResume: () {
      paused = false;
      upstream?.resume();
    },
    onCancel: () {
      stopped = true;
      final active = upstream;
      if (active != null) active.cancel().ignore();
    },
  );
  return controller.stream;
}

IdentityCheckpointFrame _decodeIdentityCheckpoint(
  $replication.IdentityCheckpoint raw,
) {
  final sequences = <String, BigInt>{};
  for (final entry in raw.lastSeqPerOrigin.entries) {
    if (!_isValidIdentityOrigin(entry.key)) {
      throw _internalSdkException('identity checkpoint has invalid origin');
    }
    sequences[entry.key] = _uint64FromFixnum(entry.value);
  }
  return IdentityCheckpointFrame._(sequences);
}

IdentityChunkFrame _decodeIdentityChunk(
  $replication.SubscribeResponse response,
) {
  if (response.writeToBuffer().length > _maxIdentityFrameBytes) {
    throw _internalSdkException('identity frame exceeds size limit');
  }
  final raw = response.identityChunk;
  final origin = _identityOriginFromBytes(raw.origin);
  final sequence = _uint64FromFixnum(raw.seq);
  if (sequence == BigInt.zero || !raw.hasHlc()) {
    throw _internalSdkException('identity chunk is missing sequence or HLC');
  }
  final hlcOrigin = _identityOriginFromBytes(raw.hlc.nodeId);
  if (hlcOrigin != origin) {
    throw _internalSdkException('identity chunk HLC origin mismatch');
  }
  final count = raw.vertexKeys.length + raw.edgeKeys.length;
  if (count > _maxIdentityChunkItems || (!raw.isLast && count == 0)) {
    throw _internalSdkException('identity chunk has invalid item count');
  }
  if (raw.chunkIndex == 0 && raw.firstItemIndex != 0) {
    throw _internalSdkException('identity chunk starts at invalid item index');
  }
  if (raw.vertexKeys.any((key) => key.isEmpty) ||
      raw.edgeKeys.any((edge) => edge.tail.isEmpty || edge.head.isEmpty)) {
    throw _internalSdkException('identity chunk has an empty graph identity');
  }
  final operation = switch (raw.operation.value) {
    1 => IdentityOperation.putVertex,
    2 => IdentityOperation.deleteVertex,
    3 => IdentityOperation.addEdge,
    4 => IdentityOperation.putEdge,
    5 => IdentityOperation.deleteEdge,
    _ => throw _internalSdkException('identity chunk has unknown operation'),
  };
  if (switch (operation) {
    IdentityOperation.putVertex ||
    IdentityOperation.deleteVertex => raw.edgeKeys.isNotEmpty,
    IdentityOperation.addEdge ||
    IdentityOperation.putEdge ||
    IdentityOperation.deleteEdge => raw.vertexKeys.isNotEmpty,
  }) {
    throw _internalSdkException(
      'identity chunk mixes operation and key family',
    );
  }
  final edges = <EdgeRef>[];
  for (final edge in raw.edgeKeys) {
    edges.add(EdgeRef(edge.tail, edge.head));
  }
  return IdentityChunkFrame._(
    origin: origin,
    sequence: sequence,
    hlc: IdentityHlc._(
      wallNanoseconds: BigInt.parse(raw.hlc.wallNs.toString()),
      logical: raw.hlc.logical,
      nodeId: hlcOrigin,
    ),
    operation: operation,
    chunkIndex: raw.chunkIndex,
    isLast: raw.isLast,
    firstItemIndex: raw.firstItemIndex,
    vertexKeys: raw.vertexKeys,
    edgeKeys: edges,
  );
}

void _validateIdentityOrigin(String value) {
  if (!_isValidIdentityOrigin(value)) {
    throw _invalidArgumentException(
      'identity origin must be a nonzero lowercase 16-byte hex ID',
    );
  }
}

bool _isValidIdentityOrigin(String value) =>
    _identityOriginPattern.hasMatch(value) &&
    value != '00000000000000000000000000000000';

String _identityOriginFromBytes(List<int> value) {
  if (value.length != 16 || value.every((byte) => byte == 0)) {
    throw _internalSdkException('identity origin must be 16 nonzero bytes');
  }
  return value.map((byte) => byte.toRadixString(16).padLeft(2, '0')).join();
}

File: sdks/node/src/changes.ts
/** Value-free, deployment-scoped change frames from identity-only Subscribe. */

import { toBinary } from "@bufbuild/protobuf";

import { InvalidArgumentError, LanternError } from "./errors.js";
import {
  IdentityOperation as PbIdentityOperation,
  SubscribeResponseSchema,
  type SubscribeResponse,
} from "./gen/graph/v1/replication_pb.js";

const MAX_UINT64 = (1n << 64n) - 1n;
const MAX_UINT32 = 0xffff_ffff;
const MAX_FRAME_BYTES = 1 << 20;
const MAX_CHUNK_ITEMS = 1024;
const ORIGIN_PATTERN = /^(?!0{32}$)[0-9a-f]{32}$/;

function validOrigin(origin: string): boolean {
  return ORIGIN_PATTERN.test(origin);
}

function originFromBytes(bytes: Uint8Array): string {
  if (bytes.length !== 16 || bytes.every((byte) => byte === 0)) {
    throw new LanternError("identity frame has invalid origin");
  }
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");
}

function checkedCursorEntries(
  sequences: Readonly<Record<string, bigint>>,
  kind: "next" | "last",
): [string, bigint][] {
  return Object.entries(sequences).map(([origin, sequence]) => {
    if (!validOrigin(origin) || typeof sequence !== "bigint") {
      throw new InvalidArgumentError(`invalid identity ${kind} cursor entry`);
    }
    if (sequence < (kind === "next" ? 1n : 0n) || sequence > MAX_UINT64) {
      throw new InvalidArgumentError(`identity ${kind} cursor sequence is outside uint64`);
    }
    return [origin, sequence];
  });
}

/** Portable per-origin cursor. Values are the NEXT expected sequence. */
export class IdentityNextCursor {
  readonly nextSequences: Readonly<Record<string, bigint>>;

  constructor(nextSequences: Readonly<Record<string, bigint>>) {
    this.nextSequences = Object.freeze(
      Object.fromEntries(checkedCursorEntries(nextSequences, "next")),
    );
    Object.freeze(this);
  }

  /** Convert a durable LAST-applied vector, rejecting uint64 exhaustion. */
  static fromLastApplied(lastSequences: Readonly<Record<string, bigint>>): IdentityNextCursor {
    const entries = checkedCursorEntries(lastSequences, "last").map(([origin, last]) => {
      if (last === MAX_UINT64) {
        throw new InvalidArgumentError("last-applied identity sequence cannot advance");
      }
      return [origin, last + 1n] as const;
    });
    return new IdentityNextCursor(Object.fromEntries(entries));
  }
}

/** One Edge identity; weights and contribution IDs are intentionally absent. */
export interface IdentityEdgeKey {
  readonly tail: string;
  readonly head: string;
}

/** Operation category of the original committed mutation. */
export type IdentityOperation = "putVertex" | "deleteVertex" | "addEdge" | "putEdge" | "deleteEdge";

/** Nanosecond-precision causal coordinate of the original mutation. */
export interface IdentityHlc {
  readonly wallNanoseconds: bigint;
  readonly logical: number;
  readonly nodeId: string;
}

/** Atomic bootstrap cutoff. Values are LAST committed sequences. */
export interface IdentityCheckpointFrame {
  readonly kind: "checkpoint";
  readonly lastSequences: Readonly<Record<string, bigint>>;
}

/** One bounded fragment of exact graph identities from a mutation. */
export interface IdentityChunkFrame {
  readonly kind: "chunk";
  readonly origin: string;
  readonly sequence: bigint;
  readonly hlc: IdentityHlc;
  readonly operation: IdentityOperation;
  readonly chunkIndex: number;
  readonly firstItemIndex: number;
  /** Advance a durable cursor only after this final chunk is applied. */
  readonly isLast: boolean;
  readonly vertexKeys: readonly string[];
  readonly edgeKeys: readonly IdentityEdgeKey[];
}

export type IdentityFrame = IdentityCheckpointFrame | IdentityChunkFrame;

/** Options for one long-lived identity stream. */
export interface IdentitySubscribeOptions {
  /** An atomic checkpoint followed by live chunks; requires an empty cursor. */
  readonly bootstrap?: boolean;
  /** Per-origin NEXT expected sequences for resume. */
  readonly cursor?: IdentityNextCursor;
  /** Explicit stream deadline. The client's default unary timeout is ignored. */
  readonly timeoutMs?: number;
}

function operationFromWire(operation: PbIdentityOperation): IdentityOperation {
  switch (operation) {
    case PbIdentityOperation.PUT_VERTEX:
      return "putVertex";
    case PbIdentityOperation.DELETE_VERTEX:
      return "deleteVertex";
    case PbIdentityOperation.ADD_EDGE:
      return "addEdge";
    case PbIdentityOperation.PUT_EDGE:
      return "putEdge";
    case PbIdentityOperation.DELETE_EDGE:
      return "deleteEdge";
    default:
      throw new LanternError("identity chunk has unknown operation");
  }
}

/** Decode one value-free frame. Exported from this internal module for tests. */
export function decodeIdentityFrame(raw: SubscribeResponse): IdentityFrame {
  let frameBytes: number;
  try {
    frameBytes = toBinary(SubscribeResponseSchema, raw).length;
  } catch (cause) {
    throw new LanternError("invalid identity frame encoding", { cause });
  }
  if (frameBytes > MAX_FRAME_BYTES) {
    throw new LanternError("identity frame exceeds size limit");
  }
  switch (raw.event.case) {
    case "checkpoint": {
      let lastSequences: Record<string, bigint>;
      try {
        lastSequences = Object.fromEntries(
          checkedCursorEntries(raw.event.value.lastSeqPerOrigin, "last"),
        );
      } catch (cause) {
        throw new LanternError("identity checkpoint has invalid cursor", { cause });
      }
      return Object.freeze({ kind: "checkpoint", lastSequences: Object.freeze(lastSequences) });
    }
    case "identityChunk": {
      const chunk = raw.event.value;
      const origin = originFromBytes(chunk.origin);
      const hlc = chunk.hlc;
      if (typeof chunk.seq !== "bigint" || chunk.seq < 1n || chunk.seq > MAX_UINT64 || !hlc) {
        throw new LanternError("identity chunk is missing sequence or HLC");
      }
      const hlcOrigin = originFromBytes(hlc.nodeId);
      if (hlcOrigin !== origin) {
        throw new LanternError("identity chunk HLC origin mismatch");
      }
      const count = chunk.vertexKeys.length + chunk.edgeKeys.length;
      if (count > MAX_CHUNK_ITEMS || (!chunk.isLast && count === 0)) {
        throw new LanternError("identity chunk has invalid item count");
      }
      if (
        !Number.isInteger(chunk.chunkIndex) ||
        chunk.chunkIndex < 0 ||
        chunk.chunkIndex > MAX_UINT32 ||
        !Number.isInteger(chunk.firstItemIndex) ||
        chunk.firstItemIndex < 0 ||
        chunk.firstItemIndex > MAX_UINT32 ||
        chunk.firstItemIndex + count > MAX_UINT32 ||
        (chunk.chunkIndex === 0 && chunk.firstItemIndex !== 0)
      ) {
        throw new LanternError("identity chunk has invalid item index");
      }
      const operation = operationFromWire(chunk.operation);
      if (
        chunk.vertexKeys.some((key) => key.length === 0) ||
        chunk.edgeKeys.some((key) => key.tail.length === 0 || key.head.length === 0)
      ) {
        throw new LanternError("identity chunk has an empty graph identity");
      }
      if (
        operation === "putVertex" || operation === "deleteVertex"
          ? chunk.edgeKeys.length !== 0
          : chunk.vertexKeys.length !== 0
      ) {
        throw new LanternError("identity chunk mixes operation and key family");
      }
      const vertexKeys = Object.freeze([...chunk.vertexKeys]);
      const edgeKeys = Object.freeze(
        chunk.edgeKeys.map((key) => Object.freeze({ tail: key.tail, head: key.head })),
      );
      return Object.freeze({
        kind: "chunk",
        origin,
        sequence: chunk.seq,
        hlc: Object.freeze({
          wallNanoseconds: hlc.wallNs,
          logical: hlc.logical,
          nodeId: hlcOrigin,
        }),
        operation,
        chunkIndex: chunk.chunkIndex,
        firstItemIndex: chunk.firstItemIndex,
        isLast: chunk.isLast,
        vertexKeys,
        edgeKeys,
      });
    }
    case "mutation":
    case undefined:
      throw new LanternError("unexpected identity stream frame");
  }
}
